### PR TITLE
--probabilities (second attempt)

### LINKED ISSUE #566: multi class oaa progressive loss
Before looking at trying to produce a reproducible example (since currently this issue is showing up experimenting on a large kaggle data set) I wanted to check on the logic being employed and make sure there is a good chance the issue I am seeing is really an issue,

The data set is multi-class (9 classes). I compiled the current VW today from github. When I build model with 

vw.exe -d train.vw --oaa 9 -b 24 --loss_function logistic -f  VW.mod

The progressive loss is exceptionally good (0.26). Should this represent the multi-clas log loss?

When predicting a test set using raw_predictions and passing through  a sigmoid I find that the predictions (probabilities) are either all equal or one class will be essentially 1 and the other classes essentially zero. The "hand computed" multi class log loss ends up being quite horrible,

vw.exe -d test.vw -t -i  modVW.mod -r raw_pred.txt 

Is this an issue already known - I did find some discussions of issues with oaa raw_prediction but sounded like it was recently patched.

### LINKED ISSUE #724: Trying to recover probabilities from command line  in daemon mode not working.
My goal is to recover the probabilities from the command line when i am in daemon mode using python.

Based on the response given  below, which  I verified it works (thanks)
http://stackoverflow.com/questions/30764810/how-do-i-get-the-raw-predictions-r-from-vowpal-wabbit-when-running-in-daemon

I do the next:

#I start my process in Daemon
                                    os.system('vw -t -i model.vw --daemon --port 26542 --link=logistic -r /dev/stdout')

Then i start sending examples with the model above in order to get probabilities.
command='echo "2 | landstar system downgraded to sell at zacks lstr |a word_count:8" | nc localhost 26542'

                                      proc = subprocess.Popen(command,shell=True,stdout=subprocess.PIPE)

then I tried to read the output
                                      proc.stdout.read()
but the above instruction returns

'1.000000\n'  

It only returns the prediction :( which is correct. I  can see how the probabilities  were created  properly in the command line but I cannot  extract them . I tried different ways but none of them worked. I only can extract the prediction, which is useless for the task than i want to do.

Could you be so kind to tell me how I can get the probabilities?

I am using mac OS X Yosemite, version 10.10.3.
VW version 7.10.2

Thank you,

ja

### PR DESCRIPTION
**Pull request by martinpopel**
This is an alternative to PR #774 (taking into account @JohnLangford 's suggestion to rather recompute the predicted label on the fly, instead of storing it in the example). I've decided to create a new PR, so I can rebase it on the current master. Original description:

This new parameter --probabilities can be used in combination with --loss_function=logistic and either --oaa or --csoaa_ldf=mc. VW predictions (-p) will be directly class probabilities (summing to one). There is no need for specifying any --link option.

This PR fixes #566 and it should also fix #724.
See #566 for a detailed discussion.

### CODE REVIEW

File: vowpalwabbit/oaa.cc [hunk 1]
```diff
@@ -85,11 +88,94 @@ void predict_or_learn(oaa& o, LEARNER::base_learner& base, example& ec)
...
+{ free(o.pred);
+  free(o.subsample_order);
+  free(o.probs);
+}
+
+// TODO: partial code duplication with multiclass.cc:finish_example
+void finish_example_probabilities(vw& all, oaa& o, example& ec)
+{ // === Compute multiclass_log_loss
+  // TODO:
+  // What to do if the correct label is unknown, i.e. (uint32_t)-1?
```
Comment (JohnLangford) [inline vowpalwabbit/oaa.cc]: If the correct label is unknown, it seems incorrect to compute any loss.

File: vowpalwabbit/csoaa.cc [hunk 2]
```diff
@@ -470,6 +470,23 @@ void do_actual_learning(ldf& data, base_learner& base)
   if (start_K > 0)
     for (size_t k=1; k<K; k++)
       LabelDict::del_example_namespaces_from_example(*data.ec_seq[k], *data.ec_seq[0], (data.all->audit || data.all->hash_inv));
+
```
Comment (JohnLangford) [inline vowpalwabbit/csoaa.cc]: I'm confused about the application to csoaa.cc.   Why does this make sense?  Predicting the expected cost given features does not have a probability semantics in general.  Why not just go through the oaa.cc code path?  This seems like to doing the same thing in two places.

File: vowpalwabbit/parse_args.cc [hunk 3]
```diff
@@ -1003,6 +1006,13 @@ void parse_reductions(vw& all)
   all.reduction_stack.push_back(bs_setup);
 
   all.l = setup_base(all);
+
+  if (all.probabilities)
+  { if (!all.vm.count("oaa") && !all.vm.count("csoaa_ldf"))
```
Comment (JohnLangford) [inline vowpalwabbit/parse_args.cc]: How can this happen?

File: vowpalwabbit/parse_args.cc [hunk 4]
```diff
@@ -1003,6 +1006,13 @@ void parse_reductions(vw& all)
...
+      THROW("--probabilities can only be used with --oaa=N or --csoaa_ldf=mc");
+    if (!all.vm.count("loss_function") || all.vm["loss_function"].as<string>() != "logistic" )
```
Comment (JohnLangford) [inline vowpalwabbit/parse_args.cc]: Shouldn't this test be in oaa.cc?

File: vowpalwabbit/csoaa.cc (hunk 2 above)
Comment (martinpopel) [inline vowpalwabbit/csoaa.cc]: I need ldf, so I cannot use `--oaa` (yes, VW is missing `--oaa_ldf`, but that's a minor problem). In `parse_args.cc` I check that the command line is `--csoaa_ldf=mc --loss_function=logisitic`, so the probabilities are well defined. See also my last comment in #566 .

File: vowpalwabbit/parse_args.cc (hunk 3 above)
Comment (martinpopel) [inline vowpalwabbit/parse_args.cc]: Simply: someone uses `--probabilities` without `--oaa` or `--csoaa_ldf`. The only other reduction I am aware of, where `--probabilities` could be used is binary classification, but here we already can use `--link=logistic`, which has the same effect of reporting probabilities. Nevertheless, so far `--probabilities` are implemented only for oaa and csoaa_ldf, so I think it is better warn users if they try to use it elsewhere.

File: vowpalwabbit/parse_args.cc (hunk 4 above)
Comment (martinpopel) [inline vowpalwabbit/parse_args.cc]: I would need to repeat the test both in oaa.cc and csoaa.cc. Should I do it?
I felt the test (just two lines of code) fits here where I am testing `all.probabilities && (!oaa || !csoaa_ldf)` (and this test obviously cannot be moved to oaa.cc/csoaa.cc).

File: vowpalwabbit/csoaa.cc (hunk 2 above)
Comment (JohnLangford) [inline vowpalwabbit/csoaa.cc]: Ok.  I smell a bad factoring in our code, but that's beyond what you are 
trying to solve.

-John

File: vowpalwabbit/parse_args.cc (hunk 4 above)
Comment (JohnLangford) [inline vowpalwabbit/parse_args.cc]: I would prefer the definition of --probabilities in oaa.cc and csoaa.cc 
and doing the checks locally.  Modularity at a small cost in extra code 
seems worthwhile.  If you can remove the probabilities flag from the 
global datastructure and just rely on the local datastructure for each, 
that seems good to.

-John

File: vowpalwabbit/example.h [hunk 5]
```diff
@@ -52,6 +52,8 @@ typedef union
 { float scalar;
   uint32_t multiclass;
   MULTILABEL::labels multilabels;
+  float* probs; // for --probabilities --oaa
```
Comment (JohnLangford) [inline vowpalwabbit/example.h]: I don't see an appropriate free().   Is there a memory leak?

File: vowpalwabbit/oaa.cc (hunk 1 above)
Comment (martinpopel) [inline vowpalwabbit/oaa.cc]: OK, but what should I do?
If I initialize `correct_class_prob = 1` (instead of 0, as it is now), I will get `multiclass_log_loss = 0` and thus effectively, I won't update `all.sd->multiclass_log_loss` nor `all.sd->holdout_multiclass_log_loss`. Should I change it this way?
I am still not sure, if I should increase `all.sd->holdout_multiclass_log_loss`.

File: vowpalwabbit/parse_args.cc (hunk 4 above)
Comment (martinpopel) [inline vowpalwabbit/parse_args.cc]: As I wrote in #774: I can move the `--probabilities` option to `oaa.cc` and `csoaa.cc` (i.e. duplicate its definition), but probably I will still need to store it inside `all` (similarly, e.q. `csoaa_rank` aka `multilabel_prediction` is stored inside `all` although it is valid only for some reductions). The problem would be that `vw -h` would not list the `--probabilities` options (see #665).
So should I really move the option `--probabilities` to oaa.cc and csoaa.cc?

File: vowpalwabbit/oaa.cc (hunk 1 above)
Comment (JohnLangford) [inline vowpalwabbit/oaa.cc]: Why is it important to have a separate loss field?  Why not just use the 
existing loss field?  Then the existing accounting mechanisms should 
work fine.

-John

File: vowpalwabbit/parse_args.cc (hunk 4 above)
Comment (JohnLangford) [inline vowpalwabbit/parse_args.cc]: Yes, more modular is preferred.  It may be impossible to remove 
all.probabilities entirely, but you should not need to do more than 
assign to it in (cs)oaa.  Instead, use a local version of the variable.  
This will make things easier when I get some time to refactor the code 
to remove all.probabilities later and deal with other label type issues 
later.

-John

File: vowpalwabbit/oaa.cc (hunk 1 above)
Comment (martinpopel) [inline vowpalwabbit/oaa.cc]: In the final resume, I want to report both zero-one loss and multiclass log loss. So I need a separate field to store both.

BTW: I tried to understand the existing accounting mechanism, but it was quite difficult as it is scattered over several files. One of the purposes of this TODO note was to check also the *existing* mechanism for possible bugs (e.g. with regards to `all.sd->weighted_unlabeled_examples` handling and unknown labels).

Comment (JohnLangford) [inline vowpalwabbit/oaa.cc]: I think we need a more general 'random variable' object to store 
statistics like this.

I think the best thing that you can do is not record any loss when there 
is no label.  Ideally, you also record the denominator so that later on 
you can observe the 0/0 situation and correctly handle it.

-John

Comment (martinpopel) [inline vowpalwabbit/oaa.cc]: I would like to implement handling of unknown labels in `--probabilities` consistently with the rest of VW, but this is surrently impossible because VW is inconsistent in this aspect -- see #847.

File: vowpalwabbit/example.h (hunk 5 above)
Comment (martinpopel) [inline vowpalwabbit/example.h]: This is just a pointer. In oaa.cc, there is both `data.probs = calloc_or_throw<float>(data.k)` and `data.probs = calloc_or_throw<float>(data.k)`. So I hope there are no memory leaks.

Why not using `v_array`? I don't need dynamically growing array. The size is fixed to k (number of classes).

Why not allocating and freeing `probs` within each example? First, I don't see a use case for this (where I would need to access `probs` of all the `ring_size` examples **simultaneously**). Second, I think allocating new memory for each example is slower than reusing one array for all the examples.

Comment (JohnLangford) [inline vowpalwabbit/example.h]: If things are to work under reduction (we want them to!), you need to have
allocated memory associated with the example rather than the reduction.
Otherwise, a reduction which calls oaa twice will get corrupted results.

-John

Comment (martinpopel) [inline vowpalwabbit/example.h]: OK, I'll change it.

**Comment by martinpopel** [comment]
`ec.pred.probs` are now allocated within each example. `all.probabilities` are now moved to oaa.cc and csoaa.cc. As for the loss of examples without labels, I am waiting for #847 to be resolved (and I would say that's a low-priority issue). Is there anything else I should fix?

File: vowpalwabbit/oaa.cc (hunk 1 above)
Comment (JohnLangford) [inline vowpalwabbit/oaa.cc]: Seems fair.

**Comment by JohnLangford** [comment]
This is merged in now.  Thanks, and thanks for your persistence.

-John

**Comment by JohnLangford** [comment]
Huh, so I thought there was a memory leak, but it turns out that _perl_ has
a memory leak.  I tweaked the test to remove perl.

-John

**Comment by martinpopel** [comment]
> Huh, so I thought there was a memory leak, but it turns out that _perl_ has a memory leak.

Out of curiosity: How have you detected the memory leak?

(Memory leaks in Perl are common when using cyclic references without weaken, but this is surely not the case of `perl -pE 's/^0/2/;s/^-1/4/'`. Even so, the leaked memory should be freed after perl exists. I would be very surprised if such simple perl command resulted in memory leaks. Have you tried sed instead of perl?)

> I tweaked the test to remove perl.

For reference: it was done in 5fe5cbfacafb44655b1c1ee83eafa25a4b6db616
I have nothing against this change (my original motivation was to prevent adding a new test file to the repository), but I think VW reading from stdin should be also tested somewhere.

**Comment by JohnLangford** [comment]
On 10/30/2015 11:21 PM, Martin Popel wrote:
>
>     Huh, so I thought there was a memory leak, but it turns out that
>     /perl/ has a memory leak.
>
> Out of curiosity: How have you detected the memory leak?
>

By prepending valgrind to the command.

> (Memory leaks in Perl are common when using cyclic references without 
> weaken, but this is surely not the case of |perl -pE 
> 's/^0/2/;s/^-1/4/'|. Even so, the leaked memory should be freed after 
> perl exists. I would be very surprised if such simple perl command 
> resulted in memory leaks. Have you tried sed instead of perl?)
>
>     I tweaked the test to remove perl.
>
> For reference: it was done in 5fe5cbf 
> <https://github.com/JohnLangford/vowpal_wabbit/commit/5fe5cbfacafb44655b1c1ee83eafa25a4b6db616>
> I have nothing against this change (my original motivation was to 
> prevent adding a new test file to the repository), but I think VW 
> reading from stdin should be also tested somewhere.
>

That's a good idea.  Maybe add a test that uses echo?

-John
